source/fundamentals/promises.txt

```diff
@@ -0,0 +1,242 @@
+from becoming unresponsive when
+executing long-running operations. For more information about asynchronous
+Javascript, see the MDN web documentation on
+`Asynchronous Javascript <https://developer.mozilla.org/en-US/docs/Learn/JavaScript/Asynchronous>`_.
+
+This section describes two features of asynchronous Javascript --
+``Promises`` and ``Callbacks`` -- that you can use with the NodeJS driver to
+access the results of your method calls to your MongoDB cluster.
+
+Promises
+--------
+
+A Promise is an object returned by the asynchronous method call that allows
+you to access information on the eventual success or failure of the operation
+that they wrap. The Promise is in the **Pending** state if the operation is
+still running, **Fulfilled** if the operation completed successfully, and
+**Rejected** if the operation threw an exception. For more information on
+Promises and related terminology, see the MDN documentation on
+:mdn:`Promises <Web/JavaScript/Reference/Global_Objects/Promise>`.
+
+Most driver methods that communicate with your MongoDB cluster such as
+``findOneAndUpdate()``, ``countDocuments()``, and ``update()`` return Promise
+objects and already contain logic to handle the success or failure of the
+operation.
+
+You can define your own logic that executes once the Promise reaches the
+**Fulfilled** or **Rejected** state by appending the ``then()`` method.
+The first parameter of ``then()`` is the method that gets called when the
+Promise reaches the **Fulfilled** state and the optional second parameter is
+the method that gets called when it reaches the **Rejected** state. The
+``then()`` method returns a Promise to which you can append additional
+``then()`` methods.
+
+When you append one or more ``then()`` methods to a Promise, each call passes
+its execution result to the next one. This pattern is called
+**Promise chaining**. The following code snippet shows an example of Promise
+chaining by appending a single ``then()`` method.
+
+.. code-block:: js
+
+   collection
+     .updateOne({ name: "Mount McKinley" }, { $set: { meters: 6190 } })
+     .then(
+       res => console.log(`Updated ${res.result.n} documents`),
+       err => console.error(`Something went wrong: ${err}`),
+     );
+
+If you only need to handle Promise transitions to the **Rejected** state,
+rather than passing a ``null`` first parameter to ``then()``, you can instead
+use the ``catch()`` method which accepts a single callback, executed when the
+Promise transitions to the **Rejected** state.
+
+The ``catch()`` method is often appended at the end of a Promise chain to
+handle any exceptions thrown. The following code snippet demonstrates appending
+a ``catch()`` method to the end of a Promise chain.
+
+.. code-block:: js
+
+   deleteOne({ name: "Mount Doom" })
+     .then(result => {
+       if (result.deletedCount !== 1) {
+         throw "Could not find Mount Doom!";
+       }
+       return incinerateNazgul();
```

Comment:
Does `incinerateNazgul()` return a promise? Is this function part of the driver?

I realize this is just an example, but to the new user or developer, this may be confusing. It is worth elaborating what this function is or return the following instead:

```
return new Promise((resolve, reject) => {
    // ...
});
```

Reply:
Would it be clear enough if I renamed the method `incinerateNazgulPromise()` or something to that effect? I wanted to focus this example on chaining Promises rather than show how to declare a Promise.

Reply:
I think that is probably fine. Or add a comment above it that says it is a function that returns a promise.